consul-lambda-registrator/extension/extension.go

```diff
@@ -0,0 +1,310 @@
+
+	// Cleanup on return. Cancel the context and close the proxy.
+	defer cancel()
+	defer func() {
+		ext.proxyMutex.Lock()
+		defer ext.proxyMutex.Unlock()
+		ext.closeProxy()
+	}()
+
+	proxyErrChan := make(chan error)
+	go ext.runProxy(ctx, proxyErrChan)
+
+	eventErrChan := make(chan error)
+	go ext.runEvents(ctx, eventErrChan)
+
+	// Block until either the proxy returns or until the event processing loop returns.
+	select {
+	case err := <-proxyErrChan:
+		return err
+	case err := <-eventErrChan:
+		return err
+	}
+}
+
+func (ext *Extension) runProxy(ctx context.Context, errChan chan error) {
+	refresh := time.NewTicker(ext.RefreshFrequency)
+	defer refresh.Stop()
+
+	pErrChan := make(chan error)
+	for {
+		select {
+		case <-ctx.Done():
+			errChan <- nil
+			return
+		case <-refresh.C:
+			err := ext.startProxy(ctx, pErrChan)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to start proxy: %w", err)
+				return
+			}
+		case err := <-pErrChan:
+			if err != nil {
+				errChan <- fmt.Errorf("proxy failed with an error: %w", err)
+				return
+			}
+		}
+	}
+}
+
+func (ext *Extension) runEvents(ctx context.Context, errChan chan error) {
+	ext.Logger.Info("processing events")
+	err := ext.Events.ProcessEvents(ctx)
+	if err != nil {
+		errChan <- fmt.Errorf("event processing failed with an error: %w", err)
+		return
+	}
+	ext.Logger.Info("event processing completed")
+	errChan <- nil
+}
+
+func (ext *Extension) getExtensionData(ctx context.Context) (structs.ExtensionData, error) {
+	trace.Enter()
+	defer trace.Exit()
+
+	var extData structs.ExtensionData
+
+	// Retrieve the data.
+	key := fmt.Sprintf("%s%s", ext.ExtensionDataPrefix, ext.service.ExtensionPath())
+	d, err := ext.Store.Get(ctx, key)
+	if err != nil {
+		return extData, fmt.Errorf("failed to get extension data for %s: %w", key, err)
+	}
+
+	// Unmarshal.
+	if err := json.Unmarshal([]byte(d), &extData); err != nil {
+		return extData, fmt.Errorf("failed to unmarshal extension data for %s: %w", key, err)
+	}
+	return extData, nil
+}
+
+// startProxy starts, or restarts, the extension's proxy server.
+// It retrieves the configuration for the proxy and if the configuration has changed
+// it closes the existing proxy server and reconfigures a new proxy server.
+// It starts the new proxy server in a separate go routine that reports any errors to the
+// caller via errChan.
+func (ext *Extension) startProxy(ctx context.Context, errChan chan error) error {
+	trace.Enter()
+	defer trace.Exit()
+
+	const errFmt = "failed to init proxy: %w"
+
+	// Get the lambda extension configuration data for this service.
+	extData, err := ext.getExtensionData(ctx)
+	if err != nil {
+		// TODO: We need to handle the case where the data used to exist but it was deleted because
+		// this Lambda function was removed from the service mesh. In that case we should shut down
+		// the proxy server so the function can't make any more outgoing calls.
+		return fmt.Errorf(errFmt, err)
+	}
```

Comment:
I wonder if we want to continue using `ext.data` when there is a transient failure and log terrible things rather than error out the extension. What do you think?